sympy/simplify/simplify.py

```diff
@@ -896,34 +896,37 @@ def __refactor(val):
     return Add(self, other)
 
 
-def product_simplify(s):
+def product_simplify(s, **kwargs):
     """Main function for Product simplification"""
     from sympy.concrete.products import Product
 
-    terms = Mul.make_args(s)
-    p_t = [] # Product Terms
-    o_t = [] # Other Terms
+    prod_terms, other = sift(Mul.make_args(s),
+                             lambda i: isinstance(i, Product), binary=True)
 
-    for term in terms:
-        if isinstance(term, Product):
-            p_t.append(term)
+    tmp_prod_terms = []
+    for term in prod_terms:
+        f = term.function.simplify(**kwargs)
+        if f.is_zero:
+            return S.Zero
```

Comment:
Thee probably needs to be some checking here:
```python
In [4]: p = Product(Product(sin(t)**2+cos(t)**2-1, (k, 0, 1))*Product(1/k, (k, 0, 1)), (n, 0, 1))

In [5]: p
Out[5]: 
       1                                               
─┬───────────┬─                                        
 │           │     1         1                         
 │           │  ─┬────┬─   ─┬──┬─                      
 │           │   │    │  1  │  │     2         2       
 │           │   │    │  ─⋅ │  │  sin (t) + cos (t) - 1
 │           │   │    │  k  │  │                       
 │           │   │    │    k = 0                       
 │           │   k = 0                                 
 │           │                                         
     n = 0                                             

In [6]: simplify(p)
Out[6]: 0

In [7]: p.doit()
Out[7]: 
                           4
    ⎛   2         2       ⎞ 
zoo⋅⎝sin (t) + cos (t) - 1⎠ 

In [8]: p.doit().simplify()
Out[8]: nan
```

Reply:
In which way should they evaluate? But of course it is possible to skip the early exit.

Reply:
The most important thing is that we don't get a definite-looking result for a case like this that is really undefined.

Reply:
I came back to this now and it turns out that the problem is actually not in this implementation (although return 0 is not a good idea).

In current master:
```
In [3]: p = Product(sin(t)**2+cos(t)**2-1, (k, 0, 1))*Product(1/k, (k, 0, 1))

In [4]: p.simplify()
Out[4]: 0

In [5]: p.doit()
Out[5]: zoo*(sin(t)**2 + cos(t)**2 - 1)**2
```

Reply:
Which seems to come from:
```
In [3]: p = 0*Product(1/k, (k, 0, 1))

In [4]: p
Out[4]: 0
```

Reply:
I think this comes from:
https://github.com/sympy/sympy/blob/a3826118a64c252bf6170a10b87d15a65728a8f4/sympy/core/mul.py#L684-L691

where especially `is_finite == False` actually should be `is_finite != True`, but that will also give some other interesting effects, like `0*x` not evaluating to `0` with default assumptions.

Reply:
Nah, that should rather be something like:
``` python
            # we know for sure the result will be 0 except the multiplicand
            # is infinity or a matrix
            if any(isinstance(c, MatrixExpr) for c in nc_part):
                return [coeff], nc_part, order_symbols
            if any(c.is_finite == False for c in c_part):
                return [S.NaN], [], order_symbols
            possibly_not_finite = [c for c in c_part if c.is_finite == None]
            return [coeff], possibly_not_finite, order_symbols
```

Reply:
Changing this leads to that the code breaks hard. Not only test failures, but test errors. Will update once everything is ran.

Reply:
Previous was wrong. Correct is probably:
``` python
            # we know for sure the result will be 0 except the multiplicand
            # is infinity or a matrix
            if any(isinstance(c, MatrixExpr) for c in nc_part):
                return [coeff], nc_part, order_symbols
            if any(c.is_finite == False for c in c_part):
                return [S.NaN], [], order_symbols
            c_part = [c for c in c_part if c.is_finite == None]
            if not c_part:
                return [coeff], [], order_symbols
```

But quite a lot of failures and `simplify` still evaluates incorrectly to 0.

Reply:
See also #17224